Project

svouse opened the pull request:
Maintains prior functionality, adds/links project class

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-devtools/45)
<!-- Reviewable:end -->

svouse committed:
added beginning of project class

svouse committed:
Merge branch 'develop' of https://github.com/google/blockly-devtools into feature_projects

svouse committed:
BlockLibraryController no longer touches tree

svouse committed:
appController listeners now use ES6

svouse committed:
more ES6 conversion, removed unused function call

svouse committed:
fleshing out project, more ES6 replacements

svouse committed:
gave appController a project w/o crash; preparing to replace blockLibraryController attribute

svouse committed:
fixed indent

svouse committed:
began adding appController-project functionality

svouse committed:
maintained functionality, added Project

svouse committed:
removed that = this

svouse committed:
removed flag

svouse committed:
upper functions were 1 space off

svouse committed:
removed commented-out code

rachel-fenichel commented:
Reviewed 6 of 6 files at r1.
Review status: 4 of 6 files reviewed at latest revision, 24 unresolved discussions.

---

*[src/app_controller.js, line 83 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGlkZD9bE8OknVTUKm:-KnGlkZEKHIN3bewr81C:b-igt9yu) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/app_controller.js#L83)):*
> ```JavaScript
>   this.view.setLibraryTree(this.blockLibraryController);
> 
>   // create project
> ```

Capitalize comments and use periods at the ends of comments.

---

*[src/app_controller.js, line 565 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGmKMchux6-vqAkPpX:-KnGmKMchux6-vqAkPpY:b-qv3lmt) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/app_controller.js#L565)):*
> ```JavaScript
>   // Update the buttons on the screen based on whether
>   // changes have been saved.
>  // var library = this.project.currentLibrary;
> ```

Delete unused code, here and below.

---

*[src/app_controller.js, line 711 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGmVLr9TmdnnULIDXT:-KnGmVLr9TmdnnULIDXU:b-9ae44i) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/app_controller.js#L711)):*
> ```JavaScript
>   // Hide and show the block library dropdown.
>   document.getElementById('modalShadow').addEventListener('click',
>       () => {
> ```

General comment: You could have saved some of these ES6 changes for a separate PR, to make this one clearer.

---

*[src/app_view.js, line 272 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGpR6Q212w3rBC9mGK:-KnGpR6Q212w3rBC9mGL:b-vprf31) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/app_view.js#L272)):*
> ```JavaScript
> 
>   /**
>    * Removes a library from the navigation tree.
> ```

Right now there's only one library that this can possible impact, right?

---

*[src/block_library_controller.js, line 75 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGpduktUMSXht62jJX:-KnGpduktUMSXht62jJY:bwa3xka) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/block_library_controller.js#L75)):*
> ```JavaScript
>  * Removes current block from Block Library and updates the save and delete
>  * buttons so that user may save block to library and but not delete.
>  * @return {Boolean} sucess of operation (for use in appController listeners)
> ```

Lowercase `boolean` https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System

And fix spelling of "success" and add a period at the end of the sentence.

When would this return false?

---

*[src/block_library_controller.js, line 112 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGq-JnmkOPUAr-AUcy:-KnGq-JnmkOPUAr-AUcz:b-ggo7e2) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/block_library_controller.js#L112)):*
> ```JavaScript
>  * Confirms with user before clearing the block library in local storage and
>  * updating the dropdown and displaying the starter block (factory_base).
>  * @return {Boolean} sucess of operation (for use in appController listeners)
> ```

Same comments as above, and on the next function.

---

*[src/block_library_controller.js, line 162 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqBbmss-QWAbp-8Cq:-KnGqBbmss-QWAbp-8Cr:b3kre) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/block_library_controller.js#L162)):*
> ```JavaScript
>   // Add select handler to the new option.
>   this.addOptionSelectHandler(blockType);
>   //this.project. FLAGX
> ```

?

---

*[src/navigation_tree.js, line 45 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqNZT8utZpN5ocnYN:-KnGqNZT8utZpN5ocnYO:b-1ob5is) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/navigation_tree.js#L45)):*
> ```JavaScript
>   }
> ```

Delete this newline.

---

*[src/navigation_tree.js, line 162 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqUIokXIrUTT-tehz:-KnGqUIokXIrUTT-tei-:b-bu0zf1) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/navigation_tree.js#L162)):*
> ```JavaScript
>    * Adds a library to the tree.
>    */
>   addLibraryNode(blockType) {
> ```

Why does this include the block type?

---

*[src/project.js, line 23 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqI1DVEe_bP-qGTun:-KnGqI1DVEe_bP-qGTuo:b-wsong2) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L23)):*
> ```JavaScript
> /**
>  * @fileoverview The Project class represents the project data structure.
>  * A project can be defined as a collection of one or more libraries along with
> ```

"A project can be defined as" implies there are other possible definitions.  Try "A project is".

---

*[src/project.js, line 24 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqagxIKxyZT2XmFsX:-KnGqagxIKxyZT2XmFsY:b-rnd4tc) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L24)):*
> ```JavaScript
>  * @fileoverview The Project class represents the project data structure.
>  * A project can be defined as a collection of one or more libraries along with
>  * any number of toolboxes and/or workspaces
> ```

Period at end of comment.

---

*[src/project.js, line 26 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqG3Lk-fskqxICkvJ:-KnGqG3Lk-fskqxICkvK:bjlanlz) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L26)):*
> ```JavaScript
>  * any number of toolboxes and/or workspaces
>  *
>  * @author sagev (Sage Vouse)
> ```

sagev@google.com

---

*[src/project.js, line 30 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqch5CJKpEeR2h2vA:-KnGqch5CJKpEeR2h2vB:brcopue) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L30)):*
> ```JavaScript
> 
>  /**
>   * @class Project aggragates libraries, toolboxes, and workspaces to form
> ```

Please flesh this out--imagine you're just starting to look at the dev tools codebase and you want to know what a "Project" is and what to expect from it.

Also, spelling on "aggregates".

---

*[src/project.js, line 35 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqxeCOVKmcm1zN1gO:-KnGqxeCOVKmcm1zN1gP:bc0c9f2) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L35)):*
> ```JavaScript
>  class Project {
>   /**
>   * Represents a user's collection of Libraries, toolboxes, and workspaces
> ```

Lowercase "Libraries" and add a period at the end of this line and the next.
All but the first line of the comment should be indented by one extra space before the `*`.

---

*[src/project.js, line 36 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGr94jYO9MBA1_RdDN:-KnGr94jYO9MBA1_RdDO:b-une0eb) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L36)):*
> ```JavaScript
>   /**
>   * Represents a user's collection of Libraries, toolboxes, and workspaces
>   * @param {string} projectName Desired name of project
> ```

What if the name is the empty string?  Will you use that anyway?

It's okay to respond with a TODO(github issue) if you want more time to figure out the correct behaviour there.

---

*[src/project.js, line 40 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGrHIHpdJrZY_UX9pR:-KnGrHIHpdJrZY_UX9pS:b-cdg416) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L40)):*
> ```JavaScript
>   */
>   constructor(projectName) {
>   	this.projectName = projectName;
> ```

Annotate all of these with type information and descriptions.

---

*[src/project.js, line 52 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGrs2wNWkfqqi2nvpm:-KnGrs2wNWkfqqi2nvpn:b-ho1obs) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L52)):*
> ```JavaScript
>    */
>    getBlockTypes() {
>    	var libraryName;
> ```

ES6 lets you do this a bit more efficiently/with fewer variable declarations: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

---

*[src/project.js, line 56 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGrLvkqejVeVrkCscU:-KnGrLvkqejVeVrkCscV:b-52l53t) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L56)):*
> ```JavaScript
>    	var libraries = Object.keys(this.libraries);
>    	var blockTypes = [];
>    	while(libraries[numLibraries]) {
> ```

Space between `while` and `(`.

---

*[src/project.js, line 61 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGrULsEYRM0w2MvTwA:-KnGrULsEYRM0w2MvTwB:bbq8se2) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L61)):*
> ```JavaScript
>    		blockTypes = blockTypes.concat(
>    			libraryController.getStoredBlockTypes());
>    	  }
> ```

Unindent by two spaces.

---

*[src/project.js, line 91 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGs7KEi2uVzspyhyQ0:-KnGs7KEi2uVzspyhyQ1:bwtx4sm) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L91)):*
> ```JavaScript
>    /**
>     * Sets the current library.
>     * @param {string} library the library to be set
> ```

If `library.name` exists, `library` is not a string.

---

*[src/project.js, line 107 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGsHijD9_wjDVif_t-:-KnGsHijD9_wjDVif_t0:b-kp6jpv) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L107)):*
> ```JavaScript
>     */
>     setCurrentToolbox(toolboxName) {
>     	this.currentToolbox = this.toolboxes[toolboxName];
> ```

What if no toolbox with that name exists?

---

*[src/project.js, line 115 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGsLgaXlNbgGsVpPoX:-KnGsLgaXlNbgGsVpPoY:b-z5l0vp) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L115)):*
> ```JavaScript
>     */
>     setCurrentWorkspace(workspaceName) {
>     	this.currentWorkspace = workspaceName;
> ```

What if no workspace with that name exists?

---

*[src/project.js, line 120 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGsRUiWSMIEvBVA7uI:-KnGsRUiWSMIEvBVA7uJ:bsabcdi) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L120)):*
> ```JavaScript
>    /**
>     * Adds a block to the project, by adding it to the current library.
>     * @param {string} blockType the block to be added
> ```

General style comment: The first word of the parameter description should be capitalized.  So this would be
`@param {string} blockType The block to be added.`

---

*[src/project.js, line 120 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGtUQRZYKNFUhGhVII:-KnGtUQRZYKNFUhGhVIJ:bbjgjs9) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L120)):*
> ```JavaScript
>    /**
>     * Adds a block to the project, by adding it to the current library.
>     * @param {string} blockType the block to be added
> ```

Is the blockType a string that's the name of the block?  "The block to be added" sounds like it's an actual block definition.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

svouse committed:
removed dead code, uncaught flag comment, and unnecessary booleans

rachel-fenichel commented:
Reviewed 2 of 2 files at r2, 4 of 5 files at r3.
Review status: 5 of 6 files reviewed at latest revision, 21 unresolved discussions.

---

*[src/block_library_controller.js, line 162 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqBbmss-QWAbp-8Cq:-KnH0BsGYGxPRt6cIKBs:b-3ax4mm) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/block_library_controller.js#L162)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

?
</blockquote></details>

I'm still confused about the this.project.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

svouse committed:
fixed author and project.js documentation

svouse committed:
removed 'this.project'

AnmAtAnm commented:
Reviewed 1 of 6 files at r1, 2 of 5 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 24 unresolved discussions.

---

*[src/project.js, line 40 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnHGaJV7iu6ts-8Y3gQ:-KnHGaJV7iu6ts-8Y3gR:b66drsx) ([raw file](https://github.com/google/blockly-devtools/blob/f747d7f3ab3c5c818a7f6cb1992a7347f1809b8b/src/project.js#L40)):*
> ```JavaScript
>     */
>     constructor(projectName) {
>     this.projectName = projectName;
> ```

Indent 2 additional spaces within the constructor.

---

*[src/project.js, line 52 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnHGtAkuP_MhpDB03GP:-KnHGtAkuP_MhpDB03GQ:b-2z19ga) ([raw file](https://github.com/google/blockly-devtools/blob/f747d7f3ab3c5c818a7f6cb1992a7347f1809b8b/src/project.js#L52)):*
> ```JavaScript
>     */
>     getBlockTypes() {
>     	var libraryName;
> ```

two space indent. No tabs. Make sure your editor configuration is adjusted to match, and that it only adjusts lines you've touched.

---

*[src/project.js, line 188 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnHHTKuEEroe3Q0whoj:-KnHHTKuEEroe3Q0whok:b-7u786c) ([raw file](https://github.com/google/blockly-devtools/blob/f747d7f3ab3c5c818a7f6cb1992a7347f1809b8b/src/project.js#L188)):*
> ```JavaScript
>       return this.currentLibrary.warnIfUnsavedChanges();
>     };
>  }
> ```

Add end-of-file newline for consistency.  Looks like there might be some stray carriage returns here, too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: all files reviewed at latest revision, 25 unresolved discussions.

---

*[src/project.js, line 186 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnHHlXmcmC2tkrd0bpR:-KnHHlXmcmC2tkrd0bpS:b6gh2ci) ([raw file](https://github.com/google/blockly-devtools/blob/f747d7f3ab3c5c818a7f6cb1992a7347f1809b8b/src/project.js#L186)):*
> ```JavaScript
>      */
>     warnIfUnsaved() {
>       return this.currentLibrary.warnIfUnsavedChanges();
> ```

Move the warn behavior to a controller.  The Project  class seems to reflect more model aspects, so it can have state about being dirty, which the controller can act on.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

svouse committed:
fixed spacing

svouse committed:
fixed indent

svouse committed:
fixed while space

svouse committed:
changed loop to ES6

svouse committed:
storing changes

svouse committed:
Merge branch 'develop' of https://github.com/google/blockly-devtools into feature_projects

svouse committed:
fixed indent

svouse committed:
Merge branch 'develop' of https://github.com/google/blockly-devtools into feature_projects

svouse committed:
fixed indents

svouse committed:
fixed comment

AnmAtAnm commented:
Reviewed 1 of 3 files at r4, 7 of 9 files at r5.
Review status: 4 of 6 files reviewed at latest revision, 32 unresolved discussions.

---

*[src/model/project.js, line 55 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLxMUXJTyIJ3vJARie:-KnLxMUXJTyIJ3vJARif:bb7ozr7) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/model/project.js#L55)):*
> ```JavaScript
>   	var libraries = Object.keys(this.libraries);
>   	var blockTypes = [];
>   	for (const libraryName of libraries) {
> ```

tabs

---

*[src/model/project.js, line 68 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLxYRqpaJgt5YSSj12:-KnLxYRqpaJgt5YSSj13:bag35z7) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/model/project.js#L68)):*
> ```JavaScript
>    */
>   getLibraryNames() {
>     return Object.keys(this.libraries);
> ```

Not to address now, but in the back of my mind I'm wondering if there might need to be some project defined order of the block libraries or other resources. Particularly in the case of the exported .js files which may have content above or below the blocks.

Similar thoughts for the other Resource types below.

---

*[src/model/project.js, line 84 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLy24bXOaSP93_emFd:-KnLy24bXOaSP93_emFe:bvkko33) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/model/project.js#L84)):*
> ```JavaScript
>    */
>   getWorkspaceNames() {
>   	return Object.keys(this.workspaces);
> ```

tabs, here and below

---

*[src/model/project.js, line 94 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLyWSoXflWwarbmUJ2:-KnLyWSoXflWwarbmUJ3:bnmmrj1) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/model/project.js#L94)):*
> ```JavaScript
>   	var libraryName = library.name;
>   	var libraries = this.getLibraryNames();
>   	if (!$.inArray(libraryName, libraries)) {
> ```

I don't see the value of jquery here.  Why not `libraries.includes(libraryName)`?

---

*[src/model/project.js, line 121 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLyujKb29Wcq3CqLAz:-KnLyujKb29Wcq3CqLB-:bv3jlrw) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/model/project.js#L121)):*
> ```JavaScript
>    */
>   addBlockToProject(blockType) {
>     //TODO: add functionality
> ```

Maybe add `console.log()` or `throw "Unimplemented: addBlockToProject"`
Similar thought for other methods below.

---

*[src/model/project.js, line 186 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLzIQMS439c4QKUm0v:-KnLzIQMS439c4QKUm0w:blasut2) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/model/project.js#L186)):*
> ```JavaScript
>     return this.currentLibrary.warnIfUnsavedChanges();
>   };
>  }
> ```

Nit: EOF Newline

---

*[src/view/app_view.js, line 278 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLzOD2_udxYVSNcgbk:-KnLzOD2_udxYVSNcgbl:b8bll81) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/view/app_view.js#L278)):*
> ```JavaScript
>   }
> ```

Nit: Remove extra newlines

---

*[src/view/navigation_tree.js, line 25 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLzSTjG9afGEl9tkwf:-KnLzSTjG9afGEl9tkwg:b-33bucs) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/view/navigation_tree.js#L25)):*
> ```JavaScript
>  * the tree view
>  *
>  * @author sagev@google.com (Sage Vouse)
> ```

You might prefer to use you GitHub id.  Your preference

---

*[src/view/navigation_tree.js, line 157 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLzY8Qu09qsK0TkIGL:-KnLzY8Qu09qsK0TkIGM:b8nvn1h) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/view/navigation_tree.js#L157)):*
> ```JavaScript
>       {'id': blockType, 'text': blockType }, 'last', null);
>   }
> ```

nit: delete extra newline.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

svouse committed:
added TODOs

svouse committed:
fixed comment, moved warning

svouse committed:
added todos, appController now has a ProjectController

svouse committed:
more TODOs, better Project comments

svouse committed:
spacing

svouse committed:
TODO spacing

svouse commented:
*[src/block_library_controller.js, line 75 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGpduktUMSXht62jJX:-KnH45M7TTihl1EdvK6t:bwyni0p) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/block_library_controller.js#L75)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Lowercase `boolean` https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System

And fix spelling of "success" and add a period at the end of the sentence.

When would this return false?

</blockquote></details>

Turns out it's wholly unnecessary

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

svouse commented:
*[src/block_library_controller.js, line 162 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqBbmss-QWAbp-8Cq:-KnH3CCRAshMQqbGR-x5:b-6l8sa5) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/block_library_controller.js#L162)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

I'm still confused about the this.project.
</blockquote></details>

Whoops; it had no place being there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

svouse committed:
comment spacing

svouse committed:
comment indent spacing

svouse committed:
getBlockTypes() tabs

svouse committed:
getToolboxNames() tabs

svouse committed:
removed jquery

svouse committed:
added throw unimplemented

svouse committed:
removed extra spaces

svouse committed:
removed extra newline

svouse committed:
constructor indent

svouse commented:
Review status: 2 of 7 files reviewed at latest revision, 36 unresolved discussions.

---

*[src/model/project.js, line 55 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLxMUXJTyIJ3vJARie:-KnMNMpcoRlmvr9GdEx7:b-886gsx) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/model/project.js#L55)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

tabs
</blockquote></details>

Fixed

---

*[src/model/project.js, line 84 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLy24bXOaSP93_emFd:-KnMPiFEATarZ3MDOVd0:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/model/project.js#L84)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

tabs, here and below
</blockquote></details>

Done.

---

*[src/model/project.js, line 94 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLyWSoXflWwarbmUJ2:-KnMPu7vznfEQbuqGp48:b-886gsx) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/model/project.js#L94)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

I don't see the value of jquery here.  Why not `libraries.includes(libraryName)`?
</blockquote></details>

Fixed

---

*[src/model/project.js, line 121 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLyujKb29Wcq3CqLAz:-KnMQlJulygl0n3-1sg4:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/model/project.js#L121)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Maybe add `console.log()` or `throw "Unimplemented: addBlockToProject"`
Similar thought for other methods below.
</blockquote></details>

Done.

---

*[src/model/project.js, line 186 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLzIQMS439c4QKUm0v:-KnMQmwNhn8Rgz7iwfWF:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/model/project.js#L186)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Nit: EOF Newline
</blockquote></details>

Done.

---

*[src/view/app_view.js, line 278 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLzOD2_udxYVSNcgbk:-KnMQpPYs8ECEGXQmGIt:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/view/app_view.js#L278)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Nit: Remove extra newlines
</blockquote></details>

Done.

---

*[src/view/navigation_tree.js, line 157 at r5](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnLzY8Qu09qsK0TkIGL:-KnMRDe68GBxX7y7suNR:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/8f5641ca775979118490bbbd234ee41c08b7c183/src/view/navigation_tree.js#L157)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

nit: delete extra newline.
</blockquote></details>

Done.

---

*[src/app_controller.js, line 83 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGlkZD9bE8OknVTUKm:-KnH3ubIOdubFpbjt9cs:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/app_controller.js#L83)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Capitalize comments and use periods at the ends of comments.
</blockquote></details>

Done.

---

*[src/app_controller.js, line 565 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGmKMchux6-vqAkPpX:-KnH3vT7ltNIPPOzEDP4:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/app_controller.js#L565)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Delete unused code, here and below.
</blockquote></details>

Done.

---

*[src/app_view.js, line 272 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGpR6Q212w3rBC9mGK:-KnH3yXva3Ffcofzc6YX:b-6hugjg) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/app_view.js#L272)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Right now there's only one library that this can possible impact, right?
</blockquote></details>

Yup; I've updated the comment for clarity with respect to current functionality 

---

*[src/navigation_tree.js, line 45 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqNZT8utZpN5ocnYN:-KnH4CKJH-L2j64ESE3X:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/navigation_tree.js#L45)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Delete this newline.
</blockquote></details>

Done.

---

*[src/navigation_tree.js, line 162 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqUIokXIrUTT-tehz:-KnL8cgOZ9ENE_jUenU8:b-6rzgwe) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/navigation_tree.js#L162)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Why does this include the block type?
</blockquote></details>

Will change to library (also just removed it for now for the sake of this PR)

---

*[src/project.js, line 23 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqI1DVEe_bP-qGTun:-KnH4QJOZZbDc7ukX4Lw:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L23)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

"A project can be defined as" implies there are other possible definitions.  Try "A project is".
</blockquote></details>

Done.

---

*[src/project.js, line 24 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqagxIKxyZT2XmFsX:-KnH4RBTDdrC5zY-cHMJ:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L24)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Period at end of comment.
</blockquote></details>

Done.

---

*[src/project.js, line 26 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqG3Lk-fskqxICkvJ:-KnH4RoAeOZAO9QPn9B_:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L26)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

sagev@google.com
</blockquote></details>

Done.

---

*[src/project.js, line 30 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqch5CJKpEeR2h2vA:-KnLA2DBE07gVW7YV9D0:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L30)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Please flesh this out--imagine you're just starting to look at the dev tools codebase and you want to know what a "Project" is and what to expect from it.

Also, spelling on "aggregates".
</blockquote></details>

Done.

---

*[src/project.js, line 35 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGqxeCOVKmcm1zN1gO:-KnL96Con7zJTNfoC2Av:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L35)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Lowercase "Libraries" and add a period at the end of this line and the next.
All but the first line of the comment should be indented by one extra space before the `*`.
</blockquote></details>

Done.

---

*[src/project.js, line 36 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGr94jYO9MBA1_RdDN:-KnLA5XFWjFJ8NyQbkab:bj80w9x) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L36)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

What if the name is the empty string?  Will you use that anyway?

It's okay to respond with a TODO(github issue) if you want more time to figure out the correct behaviour there.
</blockquote></details>

Done; created an issue for now.

---

*[src/project.js, line 40 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGrHIHpdJrZY_UX9pR:-KnLAEBYbSlecOt4gv69:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L40)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Annotate all of these with type information and descriptions.
</blockquote></details>

Done.

---

*[src/project.js, line 52 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGrs2wNWkfqqi2nvpm:-KnLtRfsae6gEXCqerxB:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L52)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

ES6 lets you do this a bit more efficiently/with fewer variable declarations: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of
</blockquote></details>

Done.

---

*[src/project.js, line 56 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGrLvkqejVeVrkCscU:-KnLldhxyAmPn8TOSfu6:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L56)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Space between `while` and `(`.
</blockquote></details>

Done.

---

*[src/project.js, line 61 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGrULsEYRM0w2MvTwA:-KnLtYLCnZRKSt9ZDpzv:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L61)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Unindent by two spaces.
</blockquote></details>

Done.

---

*[src/project.js, line 91 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGs7KEi2uVzspyhyQ0:-KnLuam0gW6Ay6HZ2WTW:b-7ql1td) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L91)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

If `library.name` exists, `library` is not a string.
</blockquote></details>

fixed

---

*[src/project.js, line 107 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGsHijD9_wjDVif_t-:-KnM2Yu4dc0-TIM1iz7J:b-886gsx) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L107)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

What if no toolbox with that name exists?
</blockquote></details>

Fixed

---

*[src/project.js, line 115 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGsLgaXlNbgGsVpPoX:-KnM2_LbAkf_BNrfpow3:b-886gsx) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L115)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

What if no workspace with that name exists?
</blockquote></details>

Fixed

---

*[src/project.js, line 120 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGsRUiWSMIEvBVA7uI:-KnM3EwW-kpKqyuLfxR_:b-886gsx) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L120)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

General style comment: The first word of the parameter description should be capitalized.  So this would be
`@param {string} blockType The block to be added.`


</blockquote></details>

Fixed

---

*[src/project.js, line 120 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGtUQRZYKNFUhGhVII:-KnM3GG-4TnqZoQ6LMaV:bgmlv4w) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/project.js#L120)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Is the blockType a string that's the name of the block?  "The block to be added" sounds like it's an actual block definition.
</blockquote></details>

Yup; fixed

---

*[src/project.js, line 40 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnHGaJV7iu6ts-8Y3gQ:-KnLA8ewHaZ5nPEIxWsO:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/f747d7f3ab3c5c818a7f6cb1992a7347f1809b8b/src/project.js#L40)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Indent 2 additional spaces within the constructor.
</blockquote></details>

Done.

---

*[src/project.js, line 52 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnHGtAkuP_MhpDB03GP:-KnLAj02ZA0WnOwHGpYb:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/f747d7f3ab3c5c818a7f6cb1992a7347f1809b8b/src/project.js#L52)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

two space indent. No tabs. Make sure your editor configuration is adjusted to match, and that it only adjusts lines you've touched.
</blockquote></details>

Done.

---

*[src/project.js, line 186 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnHHlXmcmC2tkrd0bpR:-KnM3bOIBHRGKBbPj2pE:bc3gdp5) ([raw file](https://github.com/google/blockly-devtools/blob/f747d7f3ab3c5c818a7f6cb1992a7347f1809b8b/src/project.js#L186)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Move the warn behavior to a controller.  The Project  class seems to reflect more model aspects, so it can have state about being dirty, which the controller can act on.
</blockquote></details>

Moved; also added a todo pertaining ProjectController's place (the actual warning is technically still in BlockLibraryController)

---

*[src/project.js, line 188 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnHHTKuEEroe3Q0whoj:-KnM3dylCQOD_Q3mVfnZ:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/f747d7f3ab3c5c818a7f6cb1992a7347f1809b8b/src/project.js#L188)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Add end-of-file newline for consistency.  Looks like there might be some stray carriage returns here, too.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Reviewed 1 of 3 files at r4, 1 of 9 files at r5, 4 of 5 files at r6.
Review status: 6 of 7 files reviewed at latest revision, 33 unresolved discussions.

---

*[src/model/project.js, line 50 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnMXZ43V9CXr2NeeaTa:-KnMXZ43V9CXr2NeeaTb:bixym8i) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/model/project.js#L50)):*
> ```JavaScript
>   constructor(projectName) {
>     this.projectName = projectName;
>     // Dictionary mapping {string} library names to BlockLibraryControllers.
> ```

```
/** 
 * Dictionary mapping libary names to library controllers.
 * @type {!Object<string, !BlockLibraryController>}
 */
```

Similar for the rest.  Ping me if you're tangled up figuring out the annotation for something.

---

*[src/model/project.js, line 116 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnMWG0pEbZNui_qHIkQ:-KnMWG0pEbZNui_qHIkR:bbmnur7) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/model/project.js#L116)):*
> ```JavaScript
>     var componentName = component.name;
>     if (componentArray.includes(componentName)) {
>       return true;
> ```

`return componentArray.includes(component.name)`

---

*[src/model/project.js, line 126 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnMWOGPTk9V4trjhAZU:-KnMWOGPTk9V4trjhAZV:beosjw6) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/model/project.js#L126)):*
> ```JavaScript
>    */
>   setCurrentLibrary(library) {
>   	if (this.hasComponent(library, this.getLibraryNames())) {
> ```

Nit: tabs.

---

*[src/model/project.js, line 129 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnMWQGWWntpSzC9K8LA:-KnMWQGWWntpSzC9K8LB:b-cv7mnj) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/model/project.js#L129)):*
> ```JavaScript
>   		this.currentLibrary = this.libraries[library.name];
>   	}
>     else {
> ```

put the else on the same line as the close brace of the if:
```
if (stuff) {
  // do stuff
} else {
  // do some other stuff
}
```

Same below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Reviewed 2 of 9 files at r5.
Review status: 6 of 7 files reviewed at latest revision, 22 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Reviewed 1 of 5 files at r6.
Review status: 6 of 7 files reviewed at latest revision, 24 unresolved discussions.

---

*[src/controller/project_controller.js, line 38 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-Kn_RUS5Yiviq8A6HPwD:-Kn_RUS66JlEIpzCOgZZ:bbkzzt) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/controller/project_controller.js#L38)):*
> ```JavaScript
>   //TODO #50: manage project metadata
> class ProjectController {
>   constructor(project) {
> ```

Add a JSDoc with `@param`

---

*[src/controller/project_controller.js, line 47 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-Kn_RQ6e1CuwvJPwHB_B:-Kn_RGmeVXFokr5M5IYc:b-nstk0f) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/controller/project_controller.js#L47)):*
> ```JavaScript
>    //TODO #52: move warning behavior here
>    warnIfUnsaved() {
>     return this.project.isDirty();
> ```

This references a method from a file that was deleted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 6 of 7 files reviewed at latest revision, 26 unresolved discussions.

---

*[src/view/app_view.js, line 81 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-Kn_S8MmLsgdi_Gfi7Xr:-Kn_S8MmLsgdi_Gfi7Xs:bod82c) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/view/app_view.js#L81)):*
> ```JavaScript
>     this.win.menu = this.mainMenu;
> 
>     this.navTree = null;
> ```

Declare these member variables inside class { }, with JSDoc type information.

---

*[src/view/navigation_tree.js, line 42 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-Kn_STSDvFQJMD-PQHA6:-Kn_STSDvFQJMD-PQHA7:bod82c) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/view/navigation_tree.js#L42)):*
> ```JavaScript
>     this.libraryController = libraryController;
>     var blocks = this.libraryController.getStoredBlockTypes();
>     this.makeTree(blocks);
> ```

Declare these member variables inside class { }, with JSDoc type information.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

svouse committed:
added JSDoc type information to class variables in appView, navigationTree, and Project

svouse commented:
Review status: 6 of 7 files reviewed at latest revision, 26 unresolved discussions.

---

*[src/view/app_view.js, line 81 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-Kn_S8MmLsgdi_Gfi7Xr:-Kn_azvVEX7P9RrxOP1n:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/view/app_view.js#L81)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Declare these member variables inside class { }, with JSDoc type information.
</blockquote></details>

Done.

---

*[src/view/navigation_tree.js, line 42 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-Kn_STSDvFQJMD-PQHA6:-Kn_b1NdV6_6CPchZ_mm:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/view/navigation_tree.js#L42)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Declare these member variables inside class { }, with JSDoc type information.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

svouse committed:
improved documentation

svouse committed:
added project.js stub

svouse committed:
edited stub

svouse committed:
newline in stub

AnmAtAnm commented:
Reviewed 4 of 4 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 23 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

svouse committed:
fixed has

rachel-fenichel commented:
Make a final pass to look for tabs, then LGTM

---

Reviewed 3 of 9 files at r5, 1 of 4 files at r7.
Review status: 7 of 8 files reviewed at latest revision, 23 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45#-:-Kn_ynUUcnPjl3t0crqh:b-ee85af)*
<!-- Sent from Reviewable.io -->

svouse commented:
Review status: 7 of 8 files reviewed at latest revision, 24 unresolved discussions.

---

*[src/block_library_controller.js, line 112 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnGq-JnmkOPUAr-AUcy:-Kn_zY0uK88WD_Q-Dtt6:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/ea2137ba72dd527b2714ae71996c9b13a32c8b53/src/block_library_controller.js#L112)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Same comments as above, and on the next function.
</blockquote></details>

Done.

---

*[src/controller/project_controller.js, line 38 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-Kn_RUS5Yiviq8A6HPwD:-Kn_eWv5vY8rw97o5KNF:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/controller/project_controller.js#L38)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Add a JSDoc with `@param`
</blockquote></details>

Done.

---

*[src/controller/project_controller.js, line 47 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-Kn_RQ6e1CuwvJPwHB_B:-Kn_z_53HOHVLR3atuek:bt2vr1b) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/controller/project_controller.js#L47)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

This references a method from a file that was deleted.
</blockquote></details>

Fixed.

---

*[src/model/project.js, line 50 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnMXZ43V9CXr2NeeaTa:-Kn_zaiTm63qcqHjzZhW:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/model/project.js#L50)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

```
/** 
 * Dictionary mapping libary names to library controllers.
 * @type {!Object<string, !BlockLibraryController>}
 */
```

Similar for the rest.  Ping me if you're tangled up figuring out the annotation for something.

</blockquote></details>

Done.

---

*[src/model/project.js, line 116 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnMWG0pEbZNui_qHIkQ:-Kn_zczdWVXyAj8AY2qp:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/model/project.js#L116)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

`return componentArray.includes(component.name)`
</blockquote></details>

Done.

---

*[src/model/project.js, line 129 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/45#-KnMWQGWWntpSzC9K8LA:-Kn_ze6M9HVh-pN2OS1M:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/1324f1f17fbeaa3d2b621f56a6d135d9170f3204/src/model/project.js#L129)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

put the else on the same line as the close brace of the if:
```
if (stuff) {
  // do stuff
} else {
  // do some other stuff
}
```

Same below.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45)*
<!-- Sent from Reviewable.io -->

svouse committed:
fixed tab

AnmAtAnm commented:
Ditto.

---

Review status: 6 of 8 files reviewed at latest revision, 22 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45#-:-Kna0eX2Q87v-UwEbqzs:bs4q9op)*
<!-- Sent from Reviewable.io -->

svouse committed:
fixed tabs

svouse committed:
fixed tab

svouse committed:
fixed tabs

AnmAtAnm commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 6 of 8 files reviewed at latest revision, 22 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/45#-:-KnaeI03jlswk8kbhm0W:bnfp4nl)*
<!-- Sent from Reviewable.io -->
